### DISCUSSION TITLE
Error - BACnet stack not running

### DISCUSSION THREAD

**Discussion by drakernet2** [ChristianTremblay/BAC0#482] [Q&A]
I get this error when attempting to read properties from my devices:

    Traceback (most recent call last):
      File "/home/devadmin/PycharmProjects/bacnet_scanner_bac0/main.py", line 275, in <module>
        asyncio.run(main())
      File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
        return loop.run_until_complete(main)
      File "/usr/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
        return future.result()
      File "/home/devadmin/PycharmProjects/bacnet_scanner_bac0/main.py", line 240, in main
        bacnet_device = await device_object.get_device_properties(bacnet, router_dict, device, app_logger)
      File "/home/devadmin/PycharmProjects/bacnet_scanner_bac0/device_object.py", line 70, in get_device_properties
        await bacnet.readMultiple(f"{address} device {device_id} objectName vendorName description "
      File "/home/devadmin/PycharmProjects/bacnet_scanner_bac0/.venv/lib/python3.10/site-packages/BAC0/core/io/Read.py", line 237, in readMultiple
        raise ApplicationNotStarted("BACnet stack not running - use startApp()")
    BAC0.core.io.IOExceptions.ApplicationNotStarted: BACnet stack not running - use startApp()

However, if I comment out these lines in BAC0.core.io.Read.py, the app runs fine:

    if not self._started:
       raise ApplicationNotStarted("BACnet stack not running - use startApp()")
       
I cannot find what sets "self._started" True. I'm using BAC0 version 2024.9.20.

**Comment by ChristianTremblay** [comment]
You are probably facing an asynchronous issue where the app is not yet started... without seeing the code it's hard to tell.

_started is define at the end of the initialization of the instance (Base.py line 254)

**Comment by drakernet2** [comment]
Hi, Christian:

My code definitely reaches your startApp() method, and self._started is set
to True (I set breakpoints and traced the execution). These are some log
messages when my app starts indicating that BAC0 is starting:

    2024-09-23 06:36:04,664 - INFO    | Starting Asynchronous BAC0 version
2024.09.10 (Lite)
    2024-09-23 06:36:04,664 - INFO    | Using bacpypes3 version 0.0.98
    2024-09-23 06:36:04,664 - INFO    | Use BAC0.log_level to adjust
verbosity of the app.
    2024-09-23 06:36:04,664 - INFO    | Ex. BAC0.log_level('silence') or
BAC0.log_level('error')
    2024-09-23 06:36:04,717 - INFO    | Using ip : 192.168.1.159/24 on port
47808 | broadcast : 192.168.1.255
    2024-09-23 06:36:04,734 - INFO    | Using default JSON configuration
file

At this point, my app is able to find networks and devices using await
bacnet._discover(networks='known') no problem. It is only when my app uses
Read.py to read multiple properties does it fail:

Traceback (most recent call last):
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 649, in
run_until_complete
    return future.result()
  File "/home/devadmin/PycharmProjects/bacnet_scanner_bac0/main.py", line
240, in main
    bacnet_device = await device_object.get_device_properties(bacnet,
router_dict, device, app_logger)
  File
"/home/devadmin/PycharmProjects/bacnet_scanner_bac0/device_object.py", line
70, in get_device_properties
    await bacnet.readMultiple(f"{address} device {device_id} objectName
vendorName description "
  File
"/home/devadmin/PycharmProjects/bacnet_scanner_bac0/.venv/lib/python3.10/site-packages/BAC0/core/io/Read.py",
line 237, in readMultiple
    raise ApplicationNotStarted("BACnet stack not running - use startApp()")
BAC0.core.io.IOExceptions.ApplicationNotStarted: BACnet stack not running -
use startApp()

I see how self._started is set in Base.py, but I don't see how that is set
in Read.py.

Here is how I instantiate BAC0:
    bacnet = None
    BAC0.log_level(log_file=None, stdout=None)
    try:
        async with BAC0.start(ip) as temp_bacnet:
            bacnet = temp_bacnet
    except BAC0.core.io.IOExceptions.InitializationError as e:
        print(f'Failed to start BACnet scanner. Is your specified IP
address provided valid?: {ip}')
        sys.exit()

On a separate note, I had no idea BAC0 instantiated its own BACnet device.
I did see that device listed along with the other devices my app finds on
the network. That seems like a great feature, but my app is intended only
to list the real discovered devices; is there a way to keep the BAC0
emulated device from running?

**Comment by ChristianTremblay** [comment]
It's a central piece of all the work around BAC0 and BACpypes[3]
The application **is** a BACnet device.

This device participate to the BACnet network and must be running at all time to be sure everything works. This is completely different of a standard REST API logic where you just make request to a server.

I suspect your bacnet variable is not running all the time while your script runs, causing the _started value to be False at some point.

**Comment by drakernet2** [comment]
Christian, thank you for clarifying the role of BAC0 as a BACnet device
itself -- that makes good sense to me.

Concerning self._started, I still don't know what's going on. If I set
breakpoints in Base.py and Read.py and inspect the object ids using
id(self._started), this is what I see:

In Base.py: 94244198122528
In Read.py: 94244198122496


These are two separate object instances. That seems reasonable as I don't
see how "_started" is passed from Base.py to Read.py.

**Comment by derekatkins** [comment]
@ChristianTremblay  -- Is there any update on this?  I just hit this same issue myself:
```
Traceback (most recent call last):
...
  File "BAC0\core\io\Read.py", line 237, in readMultiple
BAC0.core.io.IOExceptions.ApplicationNotStarted: BACnet stack not running - use 
startApp()
```
I use the exact same startup code as @drakernet2 did.   
ETA:  In the past couple hours while I've been working on this, I noticed that it appears that the `discover()` method appears to call the `_disconnect()` method, which is why we find `_started = False`.  From the logs:
```
2025-02-28 14:00:57,368 - INFO    | Starting Asynchronous BAC0 version 2024.09.10 (Lite)
2025-02-28 14:00:57,369 - INFO    | Using bacpypes3 version 0.0.98
2025-02-28 14:00:57,369 - INFO    | Use BAC0.log_level to adjust verbosity of the app.
2025-02-28 14:00:57,369 - INFO    | Ex. BAC0.log_level('silence') or BAC0.log_level('error')
2025-02-28 14:00:57,462 - INFO    | Using ip : 10.1.30.202/24 on port 47808 | broadcast : 10.1.30.255
2025-02-28 14:00:57,468 - INFO    | Using default JSON configuration file
2025-02-28 14:00:57,511 - INFO    | Registered as BACnet/IP App | mode normal
2025-02-28 14:00:57,513 - INFO    | Device instance (id) : 3056571
2025-02-28 14:00:57,514 - INFO    | BAC0|3056571 connected. Entering context manager.
2025-02-28 14:00:57,516 - INFO    | Stopping all tasks
2025-02-28 14:00:57,517 - INFO    | Ok all tasks stopped
2025-02-28 14:00:57,518 - INFO    | BACnet stopped
2025-02-28 14:00:57,519 - INFO    | BAC0|3056571 disconnected. Exiting context manager.
2025-02-28 14:00:59,546 - INFO    | Found those networks : {1, 2001}
```
So, basically, the library appears to be shutting down at the end of the discover() operation.
NB: I tested with the the most recent releases of BAC0 and bacpypes3 without any changes to this behavior.

Like @drakernet2 -- the code in question where it appears the library exits is:
```
    await bacnet._discover(networks='known')
```

**Comment by ChristianTremblay** [comment]
Please share your code... there is no disconnection after a discover.

**Reply by derekatkins** [reply]
@ChristianTremblay  -- The code snippet in question:
```
async def get_routing_table(my_bacnet):
    routing_table = my_bacnet.routing_table
    return routing_table

async def main(args):
    ip = args.ip
    bacnet = None
    BAC0.log_level(log_file=None, stdout=None)
    try:
        async with BAC0.start(ip) as temp_bacnet:
            bacnet = temp_bacnet
    except BAC0.core.io.IOExceptions.InitializationError as e:
        print(f'Failed to start BACnet scanner. Is your specified IP address provided valid?: {ip}')
        sys.exit()
    await bacnet._discover(networks='known')
    discovered_devices = bacnet.discoveredDevices
    routing_table = await get_routing_table(bacnet)
    for key, device in discovered_devices.items():
        bacnet_device = await device_object.get_device_properties(bacnet, router_dict, device, app_logger)

if __name__ == '__main__':
    args = parser.parse_args()
    asyncio.run(main(args))
```
In device_object.py:
```
async def get_device_properties(bacnet, router_dict, device, app_logger):
    device_id = device['object_instance'][1]
    if device_id == 3056521: # Don't include this device
        return None
    .... (create address from ipv4 or net:mac from device object)
    await bacnet.readMultiple(f"{address} device {device_id} objectName vendorName description "
                              f"systemStatus modelName firmwareRevision applicationSoftwareVersion location")
    ... 
```

**Comment by ChristianTremblay** [comment]
```python
async def get_routing_table(my_bacnet):
    routing_table = my_bacnet.routing_table
    return routing_table

async def main(args):
    ip = args.ip
    bacnet = None
    BAC0.log_level(log_file=None, stdout=None)
    try:
        async with BAC0.start(ip) as temp_bacnet:
            bacnet = temp_bacnet
            await bacnet._discover(networks='known')
            discovered_devices = bacnet.discoveredDevices
            routing_table = await get_routing_table(bacnet)
            for key, device in discovered_devices.items():
                bacnet_device = await device_object.get_device_properties(bacnet, router_dict, device, app_logger)
    except BAC0.core.io.IOExceptions.InitializationError as e:
        print(f'Failed to start BACnet scanner. Is your specified IP address provided valid?: {ip}')
        sys.exit()
if __name__ == '__main__':
    args = parser.parse_args()
    asyncio.run(main(args))
```

**Comment by ChristianTremblay** [comment]
`async with` block serves the goal of defining the network.... do the things... then disconnect.

**Reply by derekatkins** [reply]
Funny -- I just tracked it down to that block!
```
    try:
        async with BAC0.start(ip) as temp_bacnet:
            bacnet = temp_bacnet
            print(f'acquired BAC0.start, _started=: {bacnet._started}')
        print(f'after async, _started=: {bacnet._started}')
```
Results in the following output:
```
2025-02-28 16:38:24,403 - INFO    | Using default JSON configuration file
2025-02-28 16:38:24,446 - INFO    | Registered as BACnet/IP App | mode normal
2025-02-28 16:38:24,447 - INFO    | Device instance (id) : 3056609
2025-02-28 16:38:24,447 - INFO    | BAC0|3056609 connected. Entering context manager.
acquired BAC0.start, _started=: True
2025-02-28 16:38:24,451 - INFO    | Stopping all tasks
2025-02-28 16:38:24,451 - INFO    | Ok all tasks stopped
2025-02-28 16:38:24,452 - INFO    | BACnet stopped
2025-02-28 16:38:24,453 - INFO    | BAC0|3056609 disconnected. Exiting context manager.
after async, _started=: False
```

SO...  Not being a true python programmer myself (this is someone else's code that I'm just building/using) -- what do you recommend as the best way to fix / rework this??  I'd really appreciate a hand here, @ChristianTremblay -- if you don't mind?

**Reply by ChristianTremblay** [reply]
look above, I rewrote the code and put everything inside the `async with` block

**Reply by derekatkins** [reply]
Thanks, @ChristianTremblay ..  I see now what you did there.  All I can say is that there would need to be a ton more code that would need to move within that block -- I only gave you a short snippet above.  I could certainly restructure it all that way, but.... would it be safe to run it without the `async with` at all?  I.e., could i just change the code to:
```
    bacnet = None
    BAC0.log_level(log_file=None, stdout=None)
    try:
        temp_bacnet = BAC0.start(ip)
        bacnet = temp_bacnet
    except BAC0.core.io.IOExceptions.InitializationError as e:
        print(f'Failed to start BACnet scanner. Is your specified IP address provided valid?: {ip}')
        sys.exit()

    await bacnet._discover(networks='known')
    ...
```
Or would that cause other problems?  (I'll note that I've tested it this new way and the code runs, but I don't know if I may have introduced other potential future errors by doing it this way)
Thanks in advance!!

**Reply by ChristianTremblay** [reply]
The `async with` presented is one option. Mostly for simple scripts. This way, you don't have to think about disconnection details, etc...

But depending on the way you organize your code, it can also be used in more complex apps.
For example, here is an extract from a weather app I made where I'm using the async with with a while True that keeps the app running.
Other functions are calld using the "RecucringTask" class that BAC0 provides.

```python
async def my_script() -> None:
    global bacnet
    global fini
    global app
    _device_id = int(os.environ.get("BAC0_DEVICEID"))
    _port = os.environ.get("BAC0_PORT")
    _lang = make_state_text(["en", "fr"])
    _units = make_state_text(["metric", "imperial"])

    async with BAC0.start(
        deviceId=_device_id, port=_port, localObjName="BAC0_OpenWeatherMap"
    ) as weather_device:
        # async with BAC0.lite(ip='192.168.211.201/24') as bacnet:
        # await bacnet._discover()
        for prefix in PREFIXES:
            for each in _objects:
                _new_objects = each.type(
                    name=each.name.format(prefix=prefix),
                    description=each.description.format(prefix=prefix),
                    # presentValue=None,
                    properties=each.unit_state if each.unit_state else {},
                )
            if prefix != "Current":
                for each in _forecast_only_objects:
                    _new_objects = each.type(
                        name=each.name.format(prefix=prefix),
                        description=each.description.format(prefix=prefix),
                        # presentValue=None,
                        properties=each.unit_state if each.unit_state else {},
                    )
        # General Objects
        _new_objects = binary_output(
            name="Night",
            description="Based on sunrise and sunset. Can be used to control lights",
            presentValue=False,
            properties={"inactiveText": "day", "activeText": "night"},
        )
        _new_objects = multistate_value(
            name="Language",
            description="Language for requests",
            presentValue=int(os.environ.get("OWM_LANG")),
            properties={"stateText": _lang},
        )
        _new_objects = multistate_value(
            name="Measurement_System",
            description="Metric or Imperial",
            presentValue=int(os.environ.get("OWM_UNITS")),
            properties={"stateText": _units},
        )

        _new_objects = character_string(
            name="Application_Status",
            description="Will tell is something goes bad",
            presentValue="Normal",
        )
        _new_objects = character_string(
            name="City",
            description="Location of measurement",
            presentValue=os.environ.get("OWM_CITY"),
            is_commandable=True,
        )
        _new_objects = character_string(
            name="Country",
            description="Country Location of measurement",
            presentValue=os.environ.get("OWM_COUNTRY"),
            is_commandable=True,
        )
        _new_objects.add_objects_to_application(weather_device)
        app = WeatherApp(weather_device)
        task_device = RecurringTask(
            update, delay=int(os.environ.get("OWM_REFRESHRATE"))
        )
        task_device.start()
        while True:
            await asyncio.sleep(0.1)


if __name__ == "__main__":
    run(my_script, weather_device)

```

I didn't show the update function that take care of updating weather info and local variable, but you see the point.

**Reply by derekatkins** [reply]
Thanks @ChristianTremblay --
I have restructured to this way.  But now I'm seeing a strange async/future error:
```
2025-03-03 15:36:14,928 - INFO    | Registered as BACnet/IP App | mode normal
2025-03-03 15:36:14,930 - INFO    | Device instance (id) : 3057016
2025-03-03 15:36:14,930 - INFO    | BAC0|3057016 connected. Entering context manager.
2025-03-03 15:36:14,931 - INFO    | Installing recurring task Ping Task (id:131897120036112)
Task exception was never retrieved
future: <Task finished name='Task-10' coro=<IPv4DatagramServer.confirmation() done, defined at bacpypes3/ipv4/__init__.py:253> exception=DecodingError('no more packet data')>
Traceback (most recent call last):
  File "bacpypes3/ipv4/__init__.py", line 264, in confirmation
  File "bacpypes3/comm.py", line 98, in response
  File "bacpypes3/ipv4/service.py", line 115, in confirmation
  File "bacpypes3/comm.py", line 98, in response
  File "bacpypes3/ipv4/bvll.py", line 252, in confirmation
  File "bacpypes3/comm.py", line 98, in response
  File "bacpypes3/ipv4/service.py", line 333, in confirmation
  File "bacpypes3/comm.py", line 98, in response
  File "bacpypes3/netservice.py", line 476, in confirmation
  File "bacpypes3/npdu.py", line 601, in decode
  File "bacpypes3/pdu.py", line 1929, in get_short
  File "bacpypes3/pdu.py", line 1921, in get_data
bacpypes3.errors.DecodingError: no more packet data
2025-03-03 15:36:17,938 - WARNING | BAC0.scripts.Lite | Lite | BAC0.core.functions.Alias | Request timed out for what_is_network_number, no response
2025-03-03 15:36:19,940 - INFO    | Found those networks : {222}
2025-03-03 15:36:19,941 - INFO    | Discovering network 222
2025-03-03 15:36:22,953 - INFO    | Discovery done. Found 1 devices on 1 BACnet networks.
Network numbers: {222}
```
This corresponds to the following code:
```
    try:
        async with BAC0.start(ip) as bacnet:

            await bacnet._discover(networks='known')
            print(f'Network numbers: {bacnet.known_network_numbers}')
            ....
```
Any ideas?

**Reply by ChristianTremblay** [reply]
Looks like network issues... never seen this error before
